osu.Game.Rulesets.Mania/Mods/ManiaModPlayfieldCover.cs

```diff
@@ -27,22 +35,50 @@ public abstract class ManiaModPlayfieldCover : ModHidden, IApplicableToDrawableR
         [SettingSource("Coverage", "The proportion of playfield height that notes will be hidden for.")]
         public abstract BindableNumber<float> Coverage { get; }
 
+        [SettingSource("Change coverage based on combo", "Increase the coverage as combo increases.")]
+        public BindableBool ComboBasedCoverage => new BindableBool(true);
+
+        protected readonly BindableInt Combo = new BindableInt();
+
+        public override void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
+        {
+            Combo.BindTo(scoreProcessor.Combo);
+
+            // Default value of ScoreProcessor's Rank in Hidden/Fadein Mod should be SS+
+            scoreProcessor.Rank.Value = ScoreRank.XH;
+        }
+
         public virtual void ApplyToDrawableRuleset(DrawableRuleset<ManiaHitObject> drawableRuleset)
         {
             ManiaPlayfield maniaPlayfield = (ManiaPlayfield)drawableRuleset.Playfield;
 
+            if (ComboBasedCoverage.Value)
+            {
+                coveringWrappers = new List<PlayfieldCoveringWrapper>();
+                clock = maniaPlayfield.Clock;
+            }
+
             foreach (Column column in maniaPlayfield.Stages.SelectMany(stage => stage.Columns))
             {
                 HitObjectContainer hoc = column.HitObjectArea.HitObjectContainer;
                 Container hocParent = (Container)hoc.Parent;
 
                 hocParent.Remove(hoc, false);
-                hocParent.Add(new PlayfieldCoveringWrapper(hoc).With(c =>
+
+                PlayfieldCoveringWrapper coveringWrapper = new PlayfieldCoveringWrapper(hoc).With(c =>
                 {
                     c.RelativeSizeAxes = Axes.Both;
                     c.Direction = ExpandDirection;
                     c.Coverage = Coverage.Value;
-                }));
+                });
+
+                hocParent.Add(coveringWrapper);
+
+                if (ComboBasedCoverage.Value)
+                {
+                    Combo.ValueChanged += _ => updateCoverage();
```

Comment:
This `ValueChanged` callback is being registered in a loop. So it will fire N times where N is the number of the columns on the map. Given that `updateCoverage()` is not a method that would affect only this one column this is a straight-up bug causing too many updates to fire.